Adding graceful shutdown to server

nmittler opened the pull request:
Fixes https://github.com/bluejekyll/trust-dns/issues/1976.

nmittler commented:
@djc @bluejekyll this is a WIP, but I wanted to send it out to get early feedback.

nmittler commented:
@djc regarding your https://github.com/bluejekyll/trust-dns/issues/1976#issuecomment-1606108256, looks like we don't currently depend on hyper. Not sure if it would be acceptable to bring that in as a general dependency. Thoughts?

djc commented:
I did not mean bringing in hyper as a dependency, merely copying the way it implements graceful shutdown (which seems like it might be lighter-weight than the linkerd stuff that you copied)?

nmittler commented:
@djc I reworked this based on your comments regarding hyper. So far, I've only implemented UDP so we can see what it looks like.

djc commented:
This seems to add a whole bunch of abstraction on top of what's there now just to add graceful shutdown. Instead, I think we should plug graceful shutdown functionality into the existing abstractions.

nmittler commented:
@djc

> This seems to add a whole bunch of abstraction on top of what's there now just to add graceful shutdown. Instead, I think we should plug graceful shutdown functionality into the existing abstractions.

My first pass with the linkerd drain crate was just this. However, after looking into a more generic approach (as is done with hyper's Server API), I ran problems with the current structure of `ServerFuture`...

Ideally, when we start each server loop we would pass in a graceful shutdown future that can be used in `stream.take_until` to stop accepting incoming connections. However, if we pre-create the server loops when we `register` the socket, there's no easy way to insert a graceful shutdown future later on.

A few paths forward that come to mind:

1. Provide generic `registerXXX` methods.  This might be the least invasive approach. The idea here is that for every `register` method we have a generic version of the method that accepts a user-defined future for graceful shutdown. For example, we would have both ```fn register_socket(UdpSocket)``` and
```
pub fn register_socket_graceful<Fut, Out>(&mut self, socket: net::UdpSocket, drain: Fut)
    where
        Fut: Future<Output = Out> + Send + Unpin + 'static,
        Out: Send + Unpin + 'static {
    ...
    }
```
Note that because the future is internally being passed into `tokio::spawn`, both the `Future` and its `Output` have to be `static` to satisfy the lifetime requirements. 

2. Add the drain future as a member of `ServerFuture`. This actually gets tricky, since all of the generic parameters have to now be added to `ServerFuture`, itself. This makes it difficult to have a simple `new` method that takes only the handler, because there is no way to infer all of the other parameters.
3. Break apart the structure so that we don't create the server loops until later when the decision of whether to use graceful shutdown is made. In this case, the `registerXXX` methods would just add futures, and then we would have a `serve` or `run` method that takes an optional graceful shutdown future. This ends up requiring the most restructuring of the server code, since the storage of the futures requires a known size. I solved this in the new code by using a `Server` enum, and then having a `CompositeServer` that mainains a list of them.

I'm definitlely not a rust expert so there may better ways around some of the issues I was encountering. Although at the end of the day, the `Server` structure that I came up with seems pretty clean, albeit a bit of a departure from the current code.

I'm definitely open to suggestions.

@djc @bluejekyll thoughts?

bluejekyll commented:
I'm coming to this conversation a little late. Looking briefly, I'm not sure why the new server is necessary?

Thinking about how I solve similar problems,I would probably just share an `AtomicBool` to all the handler loops, and then have the loops exit cleanly after all current requests complete? I haven't put a ton of thought into this right now, but is there a reason something as simple as that wouldn't work?

I haven't looked at the code in a lot of detail right now, but based on my memory of it I think this would be possible. Is this something you considered? (I haven't looked at the Hyper implementation mentioned). The Type would be a handle that would be returned by the server, something roughly like:

```rust
struct ServerHandle {
    is_shutdown: Arc<AtomicBool>,    // if false, then handlers should continue accepting work
}

impl ServerHandle {
    /// Tell the server handlers to shutdown
    pub fn shutdown(&self) {
        is_shutdown.store(true);
    }

    /// Handler loops would read this every loop iteration, stop if true
    pub fn is_shutdown(&self) -> bool {
        is_shutdown.load()
    }
}
```

The `ServerFuture` then already has this method to await tasks finishing, so it just becomes a task of figuring out how we ire the above through all the handlers: https://docs.rs/trust-dns-server/0.22.1/trust_dns_server/server/struct.ServerFuture.html#method.block_until_done

nmittler commented:
@bluejekyll That's fair. The complexity currently is because I was attempting to allow an application to hook in their shutdown mechanism directly, similarly to what hyper does. This does introduce a lot of complexity. But I think you're right, we don't need to do that. Will update the PR shortly.

bluejekyll commented:
@djc do you agree with this approach, would be good to get that feedback before @nmittler puts in a ton of work.

nmittler commented:
@bluejekyll @djc I refactored things, PTAL when you get a moment.

nmittler commented on crates/server/src/server/server_future.rs:
```diff
@@ -642,71 +672,141 @@ impl<T: RequestHandler> ServerFuture<T> {
+    /// one will be chosen as the returned error for this future.
     pub async fn block_until_done(mut self) -> Result<(), ProtoError> {
-        let result = self.join_set.join_next().await;
+        if self.join_set.is_empty() {
```
Note to reviewers: I think this always should have blocked until all tasks are completed. With the addition of graceful shutdown, this needed to be corrected.

bluejekyll replied:
Thanks for the heads up, this looks good.

nmittler commented on crates/server/src/server/server_future.rs:
```diff
@@ -642,71 +672,141 @@ impl<T: RequestHandler> ServerFuture<T> {
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Triggers a graceful shutdown the server. All background tasks will stop accepting
+    /// new connections and the returned future will complete once all tasks have terminated.
+    pub fn shutdown(self) -> impl Future<Output = Result<(), ProtoError>> {
```
Alternatively, we could return both the future and a drain handle, to allow the application to handle them separately. I'm open to suggestions.

nmittler commented:
Also, not sure why the build is failing to find `tokio::select!` ... everything builds fine for me locally.

djc commented:
> @djc do you agree with this approach, would be good to get that feedback before @nmittler puts in a ton of work.

Yup, sounds like the right direction to me!

djc commented on crates/server/src/server/https_handler.rs:
```diff
@@ -5,11 +5,11 @@
 use bytes::{Bytes, BytesMut};
 use futures_util::lock::Mutex;
 use h2::server;
+use std::future::Future;
```
Please keep the `std` imports in a separate block.

djc commented on crates/server/src/server/quic_handler.rs:
```diff
@@ -5,10 +5,10 @@
-
 use bytes::{Bytes, BytesMut};
 use futures_util::lock::Mutex;
+use std::future::Future;
```
Here, too.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
+        out
+    }
+}
+/// A Future that is used to signal server threads to begin a graceful shutdown by
```
I think we should just skip this layer of abstraction, and just pass the `Arc<AtomicBool>` in directly. I also don't think the `Drain` naming is great in this context, I'd probably call it `shutting_down`.

nmittler replied:
This abstraction being a future allows us to use it in situtations like `take_until`, which is used for UDP. I also think the abstraction is pretty light, and doesn't complicate the API.

Regarding the naming, the term `drain` is a pretty standard term referring to the trigger that begins a graceful shutdown. Both hyper and linkerd mix `drain` and `graceful` throughout their API and docs. Calling `drain` is more concise than calling `start_graceful_shutdown`. I added a method `shutdown_gracefully` as an all-in-one helper that both starts the graceful shutdown and returns the waiting future.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
 /// Reap finished tasks from a `JoinSet`, without awaiting or blocking.
 fn reap_tasks(join_set: &mut JoinSet<()>) {
-    while FutureExt::now_or_never(join_set.join_next()).is_some() {}
+    if !join_set.is_empty() {
```
Why is this necessary?

nmittler replied:
I moved where this is called to the outside of the loops, where there may or may not be tasks in the join_set. If `FutureExt::now_or_never` is called with None, it hangs forever.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -14,7 +18,7 @@ use std::{
 #[cfg(feature = "dns-over-rustls")]
 use rustls::{Certificate, PrivateKey};
-use tokio::{net, task::JoinSet};
+use tokio::task::JoinSet;
```
Please avoid making unrelated changes in the same commit/PR, this makes it harder to review the changes you've made.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -48,25 +53,27 @@ impl<T: RequestHandler> ServerFuture<T> {
+        let (stream, stream_handle) =
             UdpStream::with_bound(socket, ([127, 255, 255, 254], 0).into());
-        //let request_stream = RequestStream::new(buf_stream, stream_handle);
+        let mut stream = stream.take_until(self.drain.clone());
```
Why do we need this change?

nmittler replied:
This just adds a line that wraps the original stream with one that will drain as we need.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Returns the drain handle used for initiating the graceful
+    /// shutdown.
+    pub fn get_drain(&self) -> Drain {
```
Why is this needed?

nmittler replied:
This allows the caller to use the API in a couple of different ways. They can either call `shutdown_gracefully`, which starts the shutdown and returns the shutdown future all in one shot. Alternatlively they can call `get_drain` and `block_until_done` separately, which allows the application to handle the two separately.

djc replied:
I suppose my question is, why would someone want to handle the two separately?

nmittler replied:
It's pretty common, actually. In fact, the Istio code does this for all of its server threads. If the application has its own shutdown hook (e.g. user clicks "x" on the application window, or some event is received on some channel), the shutdown hook can then trigger a drain of all the server threads. The code that triggers the shutdown would, in this case, be separate from the code that's waiting for the server to shutdown.

djc commented on crates/server/src/server/server_future.rs:
```diff
@@ -939,31 +1045,39 @@ mod tests {
+    use tokio::net::{TcpListener, UdpSocket};
+    use tokio::time::timeout;
 
-    #[test]
```
Why is this test no longer valuable?

nmittler replied:
It is, I just renamed it to `abort` to distinguish it from the test for `graceful_shutdown`.

djc commented:
> Also, not sure why the build is failing to find `tokio::select!` ... everything builds fine for me locally.

Probably because tokio's Cargo features aren't depended upon correctly.

nmittler commented:
@djc I resolved the issue with finding `tokio::select!` but now there is a different issue, that doesn't seem to be directly related to this PR: https://github.com/bluejekyll/trust-dns/actions/runs/5423528718/jobs/9861952221?pr=1977. Any thoughts?

djc commented:
No, you'll need to dig in and figure out what's going on.

nmittler commented:
@bluejekyll @djc I've confirmed that there is a build issue on main that is unrelated to this PR: https://github.com/bluejekyll/trust-dns/pull/1982.

Not sure if something changed recently? Perhaps one of these?

https://github.com/bluejekyll/trust-dns/pull/1978
https://github.com/bluejekyll/trust-dns/pull/1979

djc commented:
Cargo doesn't default to sticking to the locked dependencies. It's likely that the proc-macro2 crate is using some nightly feature that's newer than the nightly version we're using in CI or vice versa. Check its issue tracker?

nmittler commented:
I understand that I have a vested interest in this getting fixed, but at the end of the day I'm just contributing here. This feels more like a task for a maintainer, although I suspect @bluejekyll may be the only one since it's his repo? I'll take a look when I get some cycles.

nmittler commented:
FYI, looks like they're working on the issue: https://github.com/rust-lang/rust/issues/113152

Will see if I can throw together a temporary fix.

nmittler commented:
@bluejekyll @djc I've converted https://github.com/bluejekyll/trust-dns/pull/1982 into a fix for rust nightly

nmittler commented:
/retest

nmittler commented:
@djc I think I've addressed your comments. Let me know if there is anything else.

djc commented:
I'm mostly AFK for the next 2 weeks so am unlikely to be able to provide much feedback here.

nmittler commented:
@bluejekyll @djc I ended up going back to using linkerd's `drain-rs` for controlling the graceful shutdown. The problem I was seeing with using `AtomicBool` as a `Future` is that it never triggers the `Waker`, which can lead to it shutdown hanging indefinitely. The `drain-rs` lib seems pretty light to me ... I don't think it should bloat the server much.

Additionally, this PR seems to be having issues with code coverage (currently at ~61%). I've updated all uses of `ServerFuture` that I found in our tests with the new graceful shutdown logic. Looking at the output, it appears that most of the missed coverage is for existing code. I suspect the original code landed before the code coverage analysis was added? Can we just force merge this PR as-is? Adding coverage to `ServerFuture` seems to be beyond the scope here.

bluejekyll commented:
Ah, yes. I see the issue, needing to wake all the tasks on shutdown, I hadn't really been thinking about that, so thanks for pointing out the complexity here.

I guess the question I have, is drain-rs the simplest object type for this? I don't know much about that dependency, but in theory, any future aware channel would work just as well and avoid a new dependency, correct?

nmittler commented:
@bluejekyll 

> in theory, any future aware channel would work just as well and avoid a new dependency, correct?

Are you thinking there is something we can leverage from existing dependencies here? Or are you suggesting that we roll our own? AFAICT the `drain` library [only depends on tokio](https://github.com/linkerd/drain-rs/blob/main/Cargo.toml#L17), and it's just using channels under the covers. I think it would be preferrable to just use that if the only alternative is to roll our own. WDYT?

djc commented:
> @bluejekyll
> 
> > in theory, any future aware channel would work just as well and avoid a new dependency, correct?
> 
> Are you thinking there is something we can leverage from existing dependencies here? Or are you suggesting that we roll our own? AFAICT the `drain` library [only depends on tokio](https://github.com/linkerd/drain-rs/blob/main/Cargo.toml#L17), and it's just using channels under the covers. I think it would be preferrable to just use that if the only alternative is to roll our own. WDYT?

That sounds plausible to me.

nmittler commented:
@bluejekyll @djc I've rebased and also added a small wrapper around the linkerd drain signal, so that we don't expose their API from `ServerFuture`. PTAL

Also, any thoughts on this?

> Additionally, this PR seems to be having issues with code coverage (currently at ~61%). I've updated all uses of `ServerFuture` that I found in our tests with the new graceful shutdown logic. Looking at the output, it appears that most of the missed coverage is for existing code. I suspect the original code landed before the code coverage analysis was added? Can we just force merge this PR as-is? Adding coverage to `ServerFuture` seems to be beyond the scope here.

bluejekyll commented:
Thanks for all the explanations, @nmittler, I agree with this usage. I also just took a look at the code, seems reasonable enough to me. On the code coverage, I use that to really understand he risk we might have in regards to changes coming in, not something necessarily required for us to hit.

bluejekyll reviewed (approved):
Ok, I've reviewed. I think this overall looks good. I don't want to get pedantic on naming, but I'm not sure I would name the local variables `drain` but instead something more like `shutdown` or something that's a little closer to their usage.

@djc, can you close out some of your conversations if you don't have any open concerns?

Thanks for putting this all together @nmittler , I know this has taken a while, I appreciate you working through all the discussion points and taking the time to see this through.

nmittler commented:
Thanks @bluejekyll. FYI, I just pushed a change to use `shutdown` instead of `drain`.

nmittler commented:
@djc any other comments/concerns before we land this?

bluejekyll commented:
Merging, thanks for getting this done, @nmittler!

nmittler commented:
@bluejekyll what is the cadence for pushing new artifacts? I'd like to switch Istio over to using the graceful shutdown logic when it's available?

bluejekyll commented:
Since we're on alpha's right now, I'd be happy to publish another release today.

nmittler commented:
@bluejekyll excellent! Thanks for all your help!

theduke commented:
@nmittler thanks for implementing this!

I unfortunately couldn't dedicate more time than creating the issue.

caobug commented:
We shouldn't move ownership on shutdown_gracefully(), that complicates things.

djc commented:
Please file a new issue or PR with suggestions on what we should do instead and what your use case is (why does this complicate things).
